### PR TITLE
Refactor ERC-20 Interactions in Driver

### PR DESCRIPTION
**Pull request by nlordell**
This PR just proposes a small refactor in the `driver` to make code that uses ERC-20s easier. I also flattened the `Spender` type, since having the token tied to the spender felt a bit awkward.

It introduces a `blockchain::token::Erc20` type for fetching ERC-20 token information.

### Test Plan

Rust CI - tests continue to pass.

### CODE REVIEW

File: crates/driver/src/infra/blockchain/token.rs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+    /// Returns the [`eth::TokenAddress`] of the ERC20.
+    pub fn address(&self) -> eth::TokenAddress {
+        self.contract.address().into()
+    }
+
+    /// Fetch the ERC20 allowance for the spender. See the allowance method in
+    /// EIP-20.
+    ///
+    /// https://eips.ethereum.org/EIPS/eip-20#methods
```
Comment (MartinquaXD) [inline crates/driver/src/infra/blockchain/token.rs:23]: I guess the links above the function implementations could either be dropped or modified to link to the actual function on the website.

File: crates/driver/src/infra/blockchain/token.rs [hunk 2]
```diff
@@ -0,0 +1,85 @@
+use {super::Error, crate::domain::eth};
+
+/// An ERC-20 token.
+///
+/// https://eips.ethereum.org/EIPS/eip-20
+pub struct Erc20 {
+    contract: contracts::ERC20,
```
Comment (MartinquaXD) [inline crates/driver/src/infra/blockchain/token.rs:7]: Nice component. 👍

**Review by MartinquaXD — APPROVED** [review]

File: crates/driver/src/infra/blockchain/token.rs (hunk 1 above)
Comment (nlordell) [reply crates/driver/src/infra/blockchain/token.rs:23]: They point to the `methods` section (Which includes all ERC-20 contract method documentation). It was mostly inherited from the `allowance` function documentation.

Comment (MartinquaXD) [reply crates/driver/src/infra/blockchain/token.rs:23]: Yes. But it's also possible to link to the exact method that is getting implemented.

Comment (nlordell) [reply crates/driver/src/infra/blockchain/token.rs:23]: 🤯

**Review by fleupold — APPROVED** [review]
Looks good, but I don't see where you limit the allowance to the requested one in the code (was expecting that given the PR title)

**Comment by nlordell** [comment]
> but I don't see where you limit the allowance to the requested one in the code (was expecting that given the PR title)

Yes, sorry - I forgot to change the title. I was looking through the `driver` code and noticed we do `approve(spender, type(uint256).max)`. For some reason, I (mis)remembered that we had decided to change this behaviour in the `solver` crate - but it turns out that we didn't actually change it - so I ended up leaving the refactor I had and removed the change to set the exact approval that the solvers requested.

I updated the PR title.

File: crates/driver/src/infra/blockchain/token.rs (hunk 1 above)
Comment (nlordell) [reply crates/driver/src/infra/blockchain/token.rs]: Fixed.

**Pull request merged**